rlog/2023-09-27-gossipimprovements.mdx

```diff
@@ -0,0 +1,214 @@
+
+
+This article's primary objective is to investigate the possible choices that can empower an unstructured P2P network to achieve superior performance for the broadest set of applications. 
+We look into different constraints imposed by application-specific needs (performance goals) and investigate various choices that can augment the network's performance. 
+We explore overlay designs/freshness, peer selection approaches, message-relaying mechanisms, and resilience against adversaries/peer churn. 
+We consider GossipSub a baseline protocol to explore various possibilities and decisively commit to the ones demonstrating superior performance.
+We also discuss the current state and, where applicable, propose a strategic plan for embedding new features to the GossipSub protocol. 
+
+
+## GOAL1: Low Latency Operation
+Different applications, like blockchain, streaming, etc., impose strict time bounds on network-wide message dissemination latency. 
+A message delivered after the imposed time bounds is considered as dropped. 
+An early message delivery in applications like live streaming can further enhance the viewing quality.
+
+The properties and nature of the overlay network topology significantly impact the performance of services and applications executed on top of them. 
+Studying and devising mechanisms for better overlay design and message dissemination is paramount to achieving superior performance.
+
+Interestingly, shortest-path message delivery trees have many limitations: 
+
+1) Changing network dynamics requires a quicker and continuous readjustment of the multicast tree. 
+2) The presence of resource-constrained (bandwidth/compute, etc.) nodes in the overlay can result in congestion. 
+3) Node failure can result in partitions, making many segments unreachable. 
+4) Assuring a shortest-path tree-like structure requires a detailed view of the underlying (and continuously changing) network topology. 
+
+
+Solutions involve creating multiple random trees to add redundancy [2].
+Alternatives involve building an overlay mesh and forwarding messages through the multicast delivery tree (eager push). 
+
+Metadata is shared through the overlay links so that the nodes can ask for missing messages (lazy push or pull-based operation) through the overlay links. 
+New nodes are added from the overlay on node failure, but it requires non-faulty node selection.
+
+GossipSub uses eager push (through overlay mesh) and lazy push (through IWANT messages). 
+
+The mesh degree $D_{Low} \leq D \leq D_{High}$ is crucial in deciding message dissemination latency. 
+A smaller value for $D$ results in higher latency due to increased rounds, whereas a higher $D$ reduces latency on the cost of increased redundancy. 
+It is suggested that the average mesh degree should be $D_{avg} = \ln(N) + C$ for an optimal operation, where $N$ is the network size and C is a small constant. 

```

Comment:
from [3]

Reply:
This only true if you want to keep the latency stable as the network grows, but that will cost more bandwidth (assuming you avoid congestion)

Most networks using GossipSub won't make that trade-off, and will use a fixed D (which means constant bandwidth, increasing latency as network grows)
Also, methods used to automatically estimate the network sizes will most likely be susceptible to attacks

Speaking about attacks, sybil resistance is a big part of choosing a D, since they are directly connected
Let's say you have D = 8, and then 75% of the network does an attack
Your effective D will become 8*0.25=2, until the network recovers (recovering speed depends on the scoring system)

All of that to say, choosing a D is a complex subject, and I don't think we can sum it up in one sentence like that

Reply:
0d3dacb71ff98c6f7710c491c7312009fe76f37a